feat(client-presence): System Workspace

jason-ha opened the pull request:
1. Add infrastructure for custom System Workspace to handle internal states including ClientConnectionId to ClientSessionId.
2. Move audience support to System Workspace
3. Add `attendeeJoined` implementation
4. Add test coverage for new attendee's joining and consistency of lookup. New test cases:
```
    PresenceManager
      ✔ throws when unknown attendee is requested via `getAttendee`
      when connected
        attendee
          ✔ is announced via `attendeeJoined` when new
          already known
            ✔ is available from `getAttendee` by connection id
            ✔ is available from `getAttendee` by session id
            ✔ is available from `getAttendees`
            ✔ is NOT announced when rejoined with same connection (duplicate signal)
            ✔ is NOT announced when rejoined with different connection and current information is updated
```
5. Update general update protocol to always include client connection id -> current session id entry to ensure all updates are always working with known (already registered) session id even if a prior join or broadcast were missed.

jason-ha committed:
feat(client-presence): System Workspace
add infrastructure for custom system workspace to handle internal states including ClientConnectionId to ClientSessionId

jason-ha committed:
feat(client-presence): System Workspace
Move audience support to System Workspace
Add `attendeeJoined` implementation
Remove now unused PresenceManagerInternal

jason-ha commented:
To reviewers, it is probably easiest to review the two commits individually. The first covers item 1 in description. The second covers items 2 and 3.

RishhiB commented:
What is system workspace?

jason-ha commented:
> What is system workspace?

The Presence feature supports Workspaces where callers save and read custom Presence data and notifications. See https://github.com/microsoft/FluidFramework/blob/main/packages/framework/presence/README.md. The System Workspace is one that is internal to Presence feature. In that space Presence is itself managing data among clients. Right now, the primary data shared is map from client connection id (often just `clientId` in the codebase to an id Presence feature created - stable for lifetime of process.

jason-ha committed:
fix(client-presence): only announce new sessions
Add test coverage for new attendee's joining and consistency of lookup:
```
    PresenceManager
      ✔ throws when unknown attendee is requested via `getAttendee`
      when connected
        attendee
          ✔ is announced via `attendeeJoined` when new
          already known
            ✔ is available from `getAttendee` by connection id
            ✔ is available from `getAttendee` by session id
            ✔ is available from `getAttendees`
            ✔ is NOT announced when rejoined with same connection (duplicate signal)
            ✔ is NOT announced when rejoined with different connection and current information is updated
```

sashasimic commented on packages/framework/presence/src/presenceDatastoreManager.ts:
```diff
@@ -105,7 +103,13 @@ export class PresenceDatastoreManagerImpl implements PresenceDatastoreManager {
+		systemWorkspace: PresenceStatesEntry<PresenceStatesSchema>,
 	) {
+		// eslint-disable-next-line @typescript-eslint/consistent-type-assertions
+		this.datastore = { "system:presence": systemWorkspaceDatastore } as PresenceDatastore;
```
How does the concept of datastore here equates to runtime concept of datastore? And If I understood you right yesterday - single (runtime) datastore gets partitioned/namespaced into multiple workspaces?

jason-ha replied:
There is no direct relation to container runtime datastores. Datastore is just the appropropriate generic term for both of them.

Correct, for a single runtime there is a single over-arching datastore that includes all of the workspace "sub"-datastores.
[Presence Design Document - Datastore Structure](https://microsoft.sharepoint.com/:w:/t/FluidFramework/EZalAnQjd7NKrwAyan3kmVUBN42t6j9l9bnHJgiQvXmpzw?e=dIwtWI&nav=eyJoIjoiODkxNzQ5MDUifQ)

sashasimic replied:
Ah Ok, thank you. My suggestion here is to find some other term - data segment, data partition, data module, data block, data shard, data node. Whatever else. It's not that straightforward to explain existing datastore, so adding two different meanings to it may cause further confusion.

sashasimic commented on packages/framework/presence/src/presenceDatastoreManager.ts:
```diff
@@ -145,7 +149,7 @@ export class PresenceDatastoreManagerImpl implements PresenceDatastoreManager {
 
 	public getWorkspace<TSchema extends PresenceStatesSchema>(
-		internalWorkspaceAddress: string,
+		internalWorkspaceAddress: InternalWorkspaceAddress,
```
It looks like you are safeguarding "system" name? Can app developers retrieve any workspace? Including system? Im trying to interpret "internal" here.

jason-ha replied:
There are `WorkspaceAddress`s (public) and `InternalWorkspaceAddress`s (what is actually stored from the original customer `WorkspaceAddress`). That later address is only used internally.

sashasimic replied:
In the example, I have not seen explicit call by app developer to `getWorkspace` (or maybe I missed it). It's usually just a workspace address id that helps locate value manager. Does app developer ever need to explicitly retreive workspace object?

jason-ha replied:
"Workspace" is the concept name. In API there are two types of workspaces that customers work with: 1) states workspace and 2) notifications workspace. Those are accessed by `getStates` and `getNotifications` respectively.
In the one in-repo example we have, `getStates` is found in examples/service-clients/azure-client/external-controller/src/presence.ts. `buildDicePresence` returns `DicePresence` which is the states workspace that the example works with.
All of the test code for value managers in package uses `getStates` or `getNotifications`.

sashasimic commented on packages/framework/presence/src/systemWorkspace.ts:
```diff
@@ -0,0 +1,192 @@
+		return new Set(this.attendees.values());
+	}
+
+	public getAttendee(clientId: ClientConnectionId | ClientSessionId): ISessionClient {
```
Could you describe scenarios when you would use client connection id vs client session id?

jason-ha replied:
The system today offers plenty of client connection ids (just called clientId). Presence feature adds ClientSessionId and makes them available. You use one or the other based on what you have. For example, current service audience will provide a list of client connection ids for a user. You would use client connection id to get the session client.

sashasimic commented on packages/framework/presence/src/systemWorkspace.ts:
```diff
@@ -0,0 +1,192 @@
+		return this.selfAttendee;
+	}
+
+	private ensureAttendee(
```
It's hard to interpret "ensure" without reading implementation. But then ends up being chicken and an egg problem :) Some description would be helpful here.

sashasimic commented on packages/framework/presence/src/systemWorkspace.ts:
```diff
@@ -52,67 +80,22 @@ class SystemWorkspaceImpl implements PresenceStatesInternal, SystemWorkspace {
-		} else if (order > attendee.order) {
-			attendee.order = order;
-			attendee.currentConnectionId = currentConnectionId;
+			if (!(clientSessionId in this.datastore.clientToSessionId)) {
```
Im assuming this is the only place where local client learns about remote client's session ID? I have not looked under the hood, but Is this possibly lossy signal? If so, what happens on loss? What are the expectations between the time client (client id) shows up in the audience and its client session ID is unknown?

jason-ha replied:
Great call out. Before this change code was resilient to missing entry as any entry (which is added under a session id key) would create the session client on demand.
I think there is more metadata that I want to send with each update. We can always have "header" with each update that has the current connection id to session id entry for resiliency. (Overall resiliency patterns have not been implemented.)
Alternatively, we can restore the on-demand session client creation. I think I'll go with that for now as it covers other cases.

jason-ha replied:
On demand doesn't really work out. But sending single system workspace data for current connection is sufficient to ensure any update will be working with a known session client.

sashasimic commented:
More of a general (and more naive) question - why does app developer need to be aware of Workspace concept? Could they just get an appropriate one based on Value Manager they are requesting? Im not sure how common it would be to use NotificationManager with PresenceState or share workspace/namespace across multiple value managers.

jason-ha reviewed (commented):
> How does the concept of datastore here equates to runtime concept of datastore? And If I understood you right yesterday - single (runtime) datastore gets partitioned/namespaced into multiple workspaces? 

There is no direct relation to container runtime datastores. Datastore is just the appropropriate generic term for both of them.

Correct, for a single runtime there is a single over-arching datastore that includes all of the workspace "sub"-datastores.
https://microsoft.sharepoint.com/:w:/t/FluidFramework/EZalAnQjd7NKrwAyan3kmVUBN42t6j9l9bnHJgiQvXmpzw?e=dIwtWI&nav=eyJoIjoiODkxNzQ5MDUifQ

jason-ha committed:
docs(client-presence): add comments

jason-ha committed:
merge: 'main' into presence/system-workspace

jason-ha commented:
> More of a general (and more naive) question - why does app developer need to be aware of Workspace concept? Could they just get an appropriate one based on Value Manager they are requesting? I'm not sure how common it would be to use NotificationManager with PresenceState or share workspace/namespace across multiple value managers.

Workspaces are how we separate collections of states added for one use from collections of states for other uses. [Readme](https://github.com/microsoft/FluidFramework/blob/main/packages/framework/presence/README.md). Value managers must be named and registered with workspace.

jason-ha committed:
improvement(client-presence): session id resilience
now that session clients aren't just created on demand, every update message includes the current client connection id to session id to ensure all received states have a known session client entry.

sashasimic commented:
> > More of a general (and more naive) question - why does app developer need to be aware of Workspace concept? Could they just get an appropriate one based on Value Manager they are requesting? I'm not sure how common it would be to use NotificationManager with PresenceState or share workspace/namespace across multiple value managers.
> 
> Workspaces are how we separate collections of states added for one use from collections of states for other uses. [Readme](https://github.com/microsoft/FluidFramework/blob/main/packages/framework/presence/README.md) and [a design doc entry](https://microsoft.sharepoint.com/:w:/t/FluidFramework/EZalAnQjd7NKrwAyan3kmVUBN42t6j9l9bnHJgiQvXmpzw?e=lO6DdL&nav=eyJoIjoiMTA2NTU0OTAxNiJ9). Value managers must be named and registered with workspace.

Yes, I have read Readme. Maybe I have wrongly assumed that app developers can retrieve workspace objects? If it's just an under-the-hood concept I think we are fine here.

jason-ha commented:
> > > More of a general (and more naive) question - why does app developer need to be aware of Workspace concept? Could they just get an appropriate one based on Value Manager they are requesting? I'm not sure how common it would be to use NotificationManager with PresenceState or share workspace/namespace across multiple value managers.
> > 
> > Workspaces are how we separate collections of states added for one use from collections of states for other uses. [Readme](https://github.com/microsoft/FluidFramework/blob/main/packages/framework/presence/README.md) and [a design doc entry](https://microsoft.sharepoint.com/:w:/t/FluidFramework/EZalAnQjd7NKrwAyan3kmVUBN42t6j9l9bnHJgiQvXmpzw?e=lO6DdL&nav=eyJoIjoiMTA2NTU0OTAxNiJ9). Value managers must be named and registered with workspace.
> 
> Yes, I have read Readme. Maybe I have wrongly assumed that app developers can retrieve workspace objects? If it's just an under-the-hood concept I think we are fine here.

It isn't under the hood. Each logical use of presence states (and notifications) will carve out their own workspace. That workspace will normally be a states workspace setup via `getStates`. The object they get back named `PresenceStates` is a workspace that has properties for each of the state value managers they configured. I would just have called it `PresenceWorkspace` in API, but there was interest in having a dedicated thing for notifications; so, we have `PresenceStates` and `PresenceNotifications` which are both Presence Workspaces.
